// @nutui/nutui-taro/dist/packages/_es/CollapseItem.js
var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
/*!
* @nutui/nutui-taro v3.1.15 Fri Jan 21 2022 15:42:05 GMT+0800 (GMT+08:00)
* (c) 2021 @jdf2e.
* Released under the MIT License.
*/
import { inject, ref, reactive, computed, getCurrentInstance, watch, onMounted, toRefs, nextTick, resolveComponent, openBlock, createElementBlock, normalizeClass, createElementVNode, createBlock, createCommentVNode, renderSlot, normalizeStyle } from "vue";
import Taro, { eventCenter, getCurrentInstance as getCurrentInstance$1 } from "@tarojs/taro";
import { c as createComponent } from "./component.js";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
const { create, componentName } = createComponent("collapse-item");
const _sfc_main = create({
  props: {
    title: {
      type: String,
      default: ""
    },
    subTitle: {
      type: String,
      default: ""
    },
    disabled: {
      type: Boolean,
      default: false
    },
    name: {
      type: [Number, String],
      default: -1,
      required: true
    },
    collapseRef: {
      type: Object
    }
  },
  setup(props, ctx) {
    const collapse = inject("collapseParent");
    const conHeight = ref("auto");
    const parent = reactive(collapse);
    const classes = computed(() => {
      const prefixCls = componentName;
      return {
        [prefixCls]: true,
        [`${prefixCls}-icon`]: parent.props.icon
      };
    });
    const relation = (child) => {
      if (child.proxy) {
        parent.children.push(child.proxy);
      }
    };
    relation(getCurrentInstance());
    const proxyData = reactive({
      icon: parent.props.icon,
      iconSize: parent.props.iconSize,
      iconColor: parent.props.iconColor,
      openExpanded: null,
      iconStyle: {
        transform: "rotate(0deg)",
        marginTop: parent.props.iconHeght ? "-" + parent.props.iconHeght / 2 + "px" : "-10px"
      }
    });
    const titleIconStyle = reactive({
      titleIcon: parent.props.titleIcon,
      titleIconSize: parent.props.titleIconSize,
      titleIconColor: parent.props.titleIconColor,
      titleIconPosition: parent.props.titleIconPosition
    });
    const wrapperRef = ref(null);
    const contentRef = ref(null);
    const onTransitionEnd = () => {
      nextTick(() => {
        parent.children.forEach((item1, index) => {
          let ary = Array.from(item1.$el.children);
          ary.forEach((item2, index2) => {
            if (item2.className.includes("collapse-wrapper")) {
              item2.style.willChange = "auto";
            }
          });
        });
      });
    };
    const animation = () => {
      if (parent.props.icon && !proxyData.openExpanded) {
        proxyData.iconStyle["transform"] = "rotate(0deg)";
      } else {
        proxyData.iconStyle["transform"] = "rotate(" + parent.props.rotate + "deg)";
      }
      nextTick(() => {
        const query = Taro.getEnv() === "ALIPAY" ? my.createSelectorQuery() : Taro.createSelectorQuery();
        query.selectAll(".collapse-content").boundingClientRect();
        query.exec((res) => {
          getH(res[0]);
        });
        if (!proxyData.openExpanded) {
          onTransitionEnd();
        }
      });
    };
    const open = () => {
      proxyData.openExpanded = !proxyData.openExpanded;
      animation();
    };
    const defaultOpen = () => {
      open();
      if (parent.props.icon) {
        proxyData["iconStyle"]["transform"] = "rotate(" + parent.props.rotate + "deg)";
      }
    };
    const currentName = computed(() => props.name);
    const toggleOpen = () => {
      if (parent.props.accordion) {
        parent.children.forEach((item, index) => {
          if (currentName.value == item.name) {
            item.changeOpen(!item.openExpanded);
          } else {
            item.changeOpen(false);
            item.animation();
          }
        });
        nextTick(() => {
          parent.changeVal(currentName.value);
          animation();
        });
      } else {
        parent.changeValAry(props.name);
        open();
      }
    };
    const changeOpen = (bol) => {
      proxyData.openExpanded = bol;
    };
    const expanded = computed(() => {
      if (parent) {
        return parent.isExpanded(props.name);
      }
      return null;
    });
    watch(expanded, (value, oldValue) => {
      if (value) {
        proxyData.openExpanded = true;
      }
    });
    watch(() => {
      var _a, _b;
      return (_b = (_a = ctx == null ? void 0 : ctx.slots) == null ? void 0 : _a.default) == null ? void 0 : _b.call(_a);
    }, (vnodes) => {
      setTimeout(() => {
        getRefHeight();
      }, 300);
    });
    const getH = (list) => {
      parent.children.forEach((item1, index1) => {
        let ary = Array.from(item1.$el.children);
        let _uid = ary[1].children[0]["uid"];
        let tm = list.filter((item2) => item2.id == _uid);
        if (tm && tm.length > 0) {
          let h = tm[0]["height"];
          item1.conHeight = h;
        }
      });
    };
    const getH5 = () => {
      parent.children.forEach((item1, index1) => {
        let ary = Array.from(item1.$el.children);
        let h = ary[1].children[0]["offsetHeight"];
        item1.conHeight = h;
      });
    };
    const getRefHeight = () => {
      const query = Taro.getEnv() === "ALIPAY" ? my.createSelectorQuery() : Taro.createSelectorQuery();
      query.selectAll(".collapse-content").boundingClientRect();
      query.exec((res) => {
        if (Taro.getEnv() === "WEB") {
          getH5();
        } else {
          getH(res[0]);
        }
      });
    };
    onMounted(() => {
      const { name } = props;
      const active = parent && parent.props.active;
      if (typeof active == "number" || typeof active == "string") {
        if (name == active) {
          defaultOpen();
        }
      } else if (Object.values(active) instanceof Array) {
        const f = Object.values(active).filter((item) => item == name);
        if (f.length > 0) {
          defaultOpen();
        }
      }
      if (Taro.getEnv() === "WEB") {
        getRefHeight();
      } else {
        eventCenter.once(getCurrentInstance$1().router.onReady, () => {
          getRefHeight();
        });
      }
    });
    return __spreadProps(__spreadValues(__spreadValues(__spreadValues({
      classes
    }, toRefs(proxyData)), toRefs(parent.props)), toRefs(titleIconStyle)), {
      conHeight,
      wrapperRef,
      contentRef,
      open,
      toggleOpen,
      changeOpen,
      animation
    });
  }
});
const _hoisted_1 = { class: "collapse-title" };
const _hoisted_2 = { class: "collapse-title-value" };
const _hoisted_3 = ["innerHTML"];
const _hoisted_4 = {
  key: 0,
  class: "subTitle"
};
const _hoisted_5 = ["innerHTML"];
const _hoisted_6 = {
  class: "collapse-content",
  ref: "contentRef"
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_nut_icon = resolveComponent("nut-icon");
  return openBlock(), createElementBlock("view", {
    class: normalizeClass(_ctx.classes)
  }, [
    createElementVNode("view", {
      class: normalizeClass(["collapse-item", { "item-expanded": _ctx.openExpanded }, { "nut-collapse-item-disabled": _ctx.disabled }]),
      onClick: _cache[0] || (_cache[0] = (...args) => _ctx.toggleOpen && _ctx.toggleOpen(...args))
    }, [
      createElementVNode("view", _hoisted_1, [
        createElementVNode("view", null, [
          createElementVNode("view", _hoisted_2, [
            _ctx.titleIcon ? (openBlock(), createBlock(_component_nut_icon, {
              key: 0,
              name: _ctx.titleIcon,
              size: _ctx.titleIconSize,
              color: _ctx.titleIconColor,
              class: normalizeClass(["collapse-title-icon", _ctx.titleIconPosition == "left" ? "titleIconLeft" : "titleIconRight"])
            }, null, 8, ["name", "size", "color", "class"])) : createCommentVNode("", true),
            _ctx.$slots.mTitle ? renderSlot(_ctx.$slots, "mTitle", { key: 1 }) : (openBlock(), createElementBlock("view", {
              key: 2,
              innerHTML: _ctx.title,
              class: "collapse-icon-title"
            }, null, 8, _hoisted_3))
          ])
        ])
      ]),
      _ctx.$slots.sTitle ? (openBlock(), createElementBlock("view", _hoisted_4, [
        renderSlot(_ctx.$slots, "sTitle")
      ])) : (openBlock(), createElementBlock("view", {
        key: 1,
        innerHTML: _ctx.subTitle,
        class: "subTitle"
      }, null, 8, _hoisted_5)),
      _ctx.icon ? (openBlock(), createBlock(_component_nut_icon, {
        key: 2,
        name: _ctx.icon,
        size: _ctx.iconSize,
        color: _ctx.iconColor,
        class: normalizeClass(["collapse-icon", { "col-expanded": _ctx.openExpanded }, { "collapse-icon-disabled": _ctx.disabled }]),
        style: normalizeStyle(_ctx.iconStyle)
      }, null, 8, ["name", "size", "color", "class", "style"])) : createCommentVNode("", true)
    ], 2),
    createElementVNode("view", {
      class: normalizeClass(["collapse-wrapper", _ctx.openExpanded ? "open-style" : "close-style"]),
      ref: "wrapperRef",
      style: normalizeStyle({ height: _ctx.openExpanded ? _ctx.conHeight == "auto" ? "auto" : _ctx.conHeight + "px" : 0 })
    }, [
      createElementVNode("view", _hoisted_6, [
        renderSlot(_ctx.$slots, "default")
      ], 512)
    ], 6)
  ], 2);
}
var index_taro = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index_taro as default };
