// @nutui/nutui-taro/dist/packages/_es/index2.js
/*!
* @nutui/nutui-taro v3.1.15 Fri Jan 21 2022 15:42:05 GMT+0800 (GMT+08:00)
* (c) 2021 @jdf2e.
* Released under the MIT License.
*/
import { unref } from "vue";
function isWindow(val) {
  return val === window;
}
const useTaroRect = (elementRef, Taro) => {
  let element = unref(elementRef);
  return new Promise((resolve) => {
    if (Taro.getEnv() === "WEB") {
      if (element && element.$el) {
        element = element.$el;
      }
      if (isWindow(element)) {
        const width = element.innerWidth;
        const height = element.innerHeight;
        resolve({
          top: 0,
          left: 0,
          right: width,
          bottom: height,
          width,
          height
        });
      }
      if (element && element.getBoundingClientRect) {
        resolve(element.getBoundingClientRect());
      }
      resolve({
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        width: 0,
        height: 0
      });
    } else {
      const query = Taro.createSelectorQuery();
      query.select(`#${element.id}`).boundingClientRect();
      query.exec(function(res) {
        resolve(res[0]);
      });
    }
  });
};
export { useTaroRect as u };
