// @nutui/nutui-taro/dist/packages/_es/ShortPassword.js
/*!
* @nutui/nutui-taro v3.1.15 Fri Jan 21 2022 15:42:05 GMT+0800 (GMT+08:00)
* (c) 2021 @jdf2e.
* Released under the MIT License.
*/
import { ref, computed, watch, onMounted, resolveComponent, openBlock, createElementBlock, createVNode, withCtx, createElementVNode, toDisplayString, withDirectives, vModelText, createCommentVNode, Fragment, renderList } from "vue";
import { c as createComponent } from "./component.js";
import Taro, { eventCenter, getCurrentInstance } from "@tarojs/taro";
import { _ as _export_sfc } from "./plugin-vue_export-helper.js";
const { create } = createComponent("shortpassword");
const _sfc_main = create({
  props: {
    title: {
      type: String,
      default: "\u8BF7\u8F93\u5165\u5BC6\u7801"
    },
    desc: {
      type: String,
      default: "\u60A8\u4F7F\u7528\u4E86\u865A\u62DF\u8D44\u4EA7\uFF0C\u8BF7\u8FDB\u884C\u9A8C\u8BC1"
    },
    tips: {
      type: String,
      default: "\u5FD8\u8BB0\u5BC6\u7801"
    },
    visible: {
      type: Boolean,
      default: false
    },
    modelValue: {
      type: String,
      default: ""
    },
    errorMsg: {
      type: String,
      default: ""
    },
    noButton: {
      type: Boolean,
      default: true
    },
    closeOnClickOverlay: {
      type: Boolean,
      default: true
    },
    length: {
      type: [String, Number],
      default: 6
    }
  },
  emits: ["update:modelValue", "update:visible", "complete", "change", "ok", "tips", "close", "cancel"],
  setup(props, { emit }) {
    const realInput = ref(props.modelValue);
    const realpwd = ref();
    const comLen = computed(() => range(Number(props.length)));
    const show = ref(props.visible);
    const refRandomId = Math.random().toString(36).slice(-8);
    const randRef = ref(refRandomId);
    const isWx = ref(false);
    function sureClick() {
      emit("ok", realInput.value);
    }
    function focus() {
      let dom = "";
      console.log(123);
      if (isWx.value) {
        setTimeout(() => {
          if (!document.getElementById("nut-input-real-taro-" + randRef.value))
            return;
          dom = document.getElementById("nut-input-real-taro-" + randRef.value);
          if (!dom)
            return;
          dom.focus();
        }, 150);
      } else {
        dom = document.getElementsByClassName("nut-input-real")[0];
        let h = dom.children[0];
        h.focus();
      }
    }
    watch(() => props.visible, (value) => {
      show.value = value;
      if (value) {
        randRef.value = Math.random().toString(36).slice(-8);
        if (Taro.getEnv() === "WEB") {
          isWx.value = false;
        } else {
          isWx.value = true;
        }
      }
    });
    watch(() => props.modelValue, (value) => {
      realInput.value = value;
    });
    function changeValue(e) {
      const input = e.target;
      let val = input.value;
      if (val.length > comLen.value) {
        val = val.slice(0, comLen.value);
        realInput.value = val;
      }
      if (String(realInput.value).length === comLen.value) {
        emit("complete", val);
      }
      emit("change", val);
      emit("update:modelValue", val);
    }
    function close() {
      emit("update:visible", false);
      emit("cancel");
    }
    function closeIcon() {
      emit("update:visible", false);
      emit("close");
    }
    function range(val) {
      return Math.min(Math.max(4, val), 6);
    }
    function onTips() {
      emit("tips");
    }
    onMounted(() => {
      eventCenter.once(getCurrentInstance().router.onReady, () => {
      });
      if (Taro.getEnv() === "WEB") {
        isWx.value = false;
      } else {
        isWx.value = true;
      }
    });
    return {
      comLen,
      sureClick,
      realInput,
      realpwd,
      range,
      changeValue,
      close,
      onTips,
      focus,
      show,
      closeIcon,
      isWx,
      refRandomId,
      randRef
    };
  }
});
const _hoisted_1 = { class: "nut-shortpsd-title" };
const _hoisted_2 = { class: "nut-shortpsdWx-subtitle" };
const _hoisted_3 = ["id", "maxlength"];
const _hoisted_4 = { class: "nut-input-w" };
const _hoisted_5 = /* @__PURE__ */ createElementVNode("view", { class: "nut-input-site" }, null, -1);
const _hoisted_6 = {
  key: 0,
  class: "nut-shortpsd-icon"
};
const _hoisted_7 = { class: "nut-shortpsd-message" };
const _hoisted_8 = { class: "nut-shortpsd-error" };
const _hoisted_9 = {
  key: 1,
  class: "nut-shortpsd-footer"
};
function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_nut_icon = resolveComponent("nut-icon");
  const _component_nut_popup = resolveComponent("nut-popup");
  return openBlock(), createElementBlock("view", null, [
    createVNode(_component_nut_popup, {
      style: {
        padding: "32px 24px 28px 24px",
        borderRadius: "12px",
        textAlign: "center"
      },
      visible: _ctx.show,
      "onUpdate:visible": _cache[8] || (_cache[8] = ($event) => _ctx.show = $event),
      closeable: true,
      onClickCloseIcon: _ctx.closeIcon,
      "close-on-click-overlay": _ctx.closeOnClickOverlay,
      onClickOverlay: _ctx.close
    }, {
      default: withCtx(() => [
        createElementVNode("view", _hoisted_1, toDisplayString(_ctx.title), 1),
        createElementVNode("view", _hoisted_2, toDisplayString(_ctx.desc), 1),
        _ctx.isWx && _ctx.visible ? withDirectives((openBlock(), createElementBlock("input", {
          key: 0,
          class: "nut-input-real-taro",
          id: "nut-input-real-taro-" + _ctx.randRef,
          type: "number",
          maxlength: _ctx.length,
          "onUpdate:modelValue": _cache[0] || (_cache[0] = ($event) => _ctx.realInput = $event),
          onInput: _cache[1] || (_cache[1] = (...args) => _ctx.changeValue && _ctx.changeValue(...args))
        }, null, 40, _hoisted_3)), [
          [vModelText, _ctx.realInput]
        ]) : createCommentVNode("", true),
        createElementVNode("view", _hoisted_4, [
          !_ctx.isWx ? withDirectives((openBlock(), createElementBlock("input", {
            key: 0,
            ref: "realpwd",
            class: "nut-input-real",
            type: "number",
            maxlength: "6",
            "onUpdate:modelValue": _cache[2] || (_cache[2] = ($event) => _ctx.realInput = $event),
            onInput: _cache[3] || (_cache[3] = (...args) => _ctx.changeValue && _ctx.changeValue(...args))
          }, null, 544)), [
            [vModelText, _ctx.realInput]
          ]) : createCommentVNode("", true),
          _hoisted_5,
          createElementVNode("view", {
            class: "nut-shortpsd-fake-taro",
            onClick: _cache[4] || (_cache[4] = (...args) => _ctx.focus && _ctx.focus(...args))
          }, [
            (openBlock(true), createElementBlock(Fragment, null, renderList(new Array(_ctx.comLen), (sublen, index) => {
              return openBlock(), createElementBlock("view", {
                class: "nut-shortpsd-li",
                key: index
              }, [
                String(_ctx.realInput).length > index ? (openBlock(), createElementBlock("view", _hoisted_6)) : createCommentVNode("", true)
              ]);
            }), 128))
          ])
        ]),
        createElementVNode("view", _hoisted_7, [
          createElementVNode("view", _hoisted_8, toDisplayString(_ctx.errorMsg), 1),
          _ctx.tips ? (openBlock(), createElementBlock("view", {
            key: 0,
            class: "nut-shortpsd-forget",
            onClick: _cache[5] || (_cache[5] = (...args) => _ctx.onTips && _ctx.onTips(...args))
          }, [
            createVNode(_component_nut_icon, {
              class: "icon",
              size: "11px",
              name: "tips"
            }),
            createElementVNode("view", null, toDisplayString(_ctx.tips), 1)
          ])) : createCommentVNode("", true)
        ]),
        !_ctx.noButton ? (openBlock(), createElementBlock("view", _hoisted_9, [
          createElementVNode("view", {
            class: "nut-shortpsd-cancle",
            onClick: _cache[6] || (_cache[6] = (...args) => _ctx.close && _ctx.close(...args))
          }, "\u53D6\u6D88"),
          createElementVNode("view", {
            class: "nut-shortpsd-sure",
            onClick: _cache[7] || (_cache[7] = (...args) => _ctx.sureClick && _ctx.sureClick(...args))
          }, "\u786E\u8BA4")
        ])) : createCommentVNode("", true)
      ]),
      _: 1
    }, 8, ["visible", "onClickCloseIcon", "close-on-click-overlay", "onClickOverlay"])
  ]);
}
var index_taro = /* @__PURE__ */ _export_sfc(_sfc_main, [["render", _sfc_render]]);
export { index_taro as default };
